pallets/moonbeam-foreign-assets/src/lib.rs

```diff
@@ -0,0 +1,513 @@
+	#[pallet::storage]
+	#[pallet::getter(fn assets_by_id)]
+	pub type AssetsById<T: Config> =
+		CountedStorageMap<_, Blake2_128Concat, AssetId, Location, OptionQuery>;
+
+	/// Reverse mapping of AssetsById. Mapping from a foreign asset to an asset id.
+	/// This is mostly used when receiving a multilocation XCM message to retrieve
+	/// the corresponding asset in which tokens should me minted.
+	#[pallet::storage]
+	#[pallet::getter(fn assets_by_location)]
+	pub type AssetsByLocation<T: Config> =
+		StorageMap<_, Blake2_128Concat, Location, (AssetId, AssetStatus)>;
+
+	impl<T: Config> Pallet<T> {
+		/// The account ID of this pallet
+		#[inline]
+		pub fn account_id() -> H160 {
+			let account_id: T::AccountId = PALLET_ID.into_account_truncating();
+			T::AccountIdToH160::convert(account_id)
+		}
+
+		/// Compute asset contract address from asset id
+		#[inline]
+		pub(crate) fn contract_address_from_asset_id(asset_id: AssetId) -> H160 {
+			let mut buffer = [0u8; 20];
+			buffer[..4].copy_from_slice(&FOREIGN_ASSETS_PREFIX);
+			buffer[4..].copy_from_slice(&asset_id.to_be_bytes());
+			H160(buffer)
+		}
+
+		/// Mint an asset into a specific account
+		pub fn mint_into(
+			asset_id: AssetId,
+			beneficiary: T::AccountId,
+			amount: U256,
+		) -> Result<(), evm::EvmError> {
+			// We perform the evm call in a storage transaction to ensure that if it fail
+			// any contract storage changes are rolled back.
+			frame_support::storage::with_storage_layer(|| {
+				EvmCaller::<T>::erc20_mint_into(
+					Self::contract_address_from_asset_id(asset_id),
+					T::AccountIdToH160::convert(beneficiary),
+					amount,
+				)
+			})
+			.map_err(Into::into)
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Create new asset with the ForeignAssetCreator
+		#[pallet::call_index(0)]
+		#[pallet::weight(<T as Config>::WeightInfo::create_foreign_asset())]
+		pub fn create_foreign_asset(
+			origin: OriginFor<T>,
+			asset_id: AssetId,
+			xcm_location: Location,
+			decimals: u8,
+			symbol: BoundedVec<u8, ConstU32<256>>,
+			name: BoundedVec<u8, ConstU32<256>>,
+		) -> DispatchResult {
+			T::ForeignAssetCreatorOrigin::ensure_origin(origin)?;
+
+			// Ensure such an assetId does not exist
+			ensure!(
+				!AssetsById::<T>::contains_key(&asset_id),
+				Error::<T>::AssetAlreadyExists
+			);
+
+			ensure!(
+				!AssetsByLocation::<T>::contains_key(&xcm_location),
+				Error::<T>::LocationAlreadyExists
+			);
+
+			ensure!(
+				AssetsById::<T>::count() < T::MaxForeignAssets::get(),
+				Error::<T>::TooManyForeignAssets
+			);
+
+			ensure!(
+				T::AssetIdFilter::contains(&asset_id),
+				Error::<T>::AssetIdFiltered
+			);
+
+			let symbol = core::str::from_utf8(&symbol).map_err(|_| Error::<T>::InvalidTicker)?;
+			let name = core::str::from_utf8(&name).map_err(|_| Error::<T>::InvalidTokenName)?;
+
+			let contract_address = EvmCaller::<T>::erc20_create(asset_id, decimals, symbol, name)?;
```

Comment:
Shouldn't we wrap these calls inside a `with_storage_layer` as well? I guess that the answer to that is "no", given that we are already in a context of a pallet call and any storage changes are rolled back if the call fails, right?

Reply:
Yes, we need `with_storage_layer` only for automatic execution (=anything outside of the context of an extrinsic)